core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java

```diff
@@ -65,431 +59,179 @@ public ZooKeeper getZooKeeper() {
-                return;
-              }
-              throw e;
-            }
-            // Let other KeeperException bubble to the outer catch
-          } else {
-            // If the stat is null, the node is now gone which is fine.
-            return;
-          }
-        } catch (KeeperException e) {
-          final Code c = e.code();
-          if (c == Code.CONNECTIONLOSS || c == Code.OPERATIONTIMEOUT || c == Code.SESSIONEXPIRED) {
-            retryOrThrow(retry, e);
-          } else {
-            throw e;
-          }
-        }
-
-        retry.waitForNextAttempt();
-      }
-    } catch (KeeperException e) {
-      if (policy.equals(NodeMissingPolicy.SKIP) && e.code().equals(KeeperException.Code.NONODE))
-        return;
-      throw e;
-    }
+    return retryLoop(zk -> zk.getACL(zPath, stat));
   }
 
   /**
    * Create a persistent node with the default ACL
-   *
-   * @return true if the node was created or altered; false if it was skipped
    */
-  public boolean putPersistentData(String zPath, byte[] data, NodeExistsPolicy policy)
+  public void putPersistentData(String zPath, byte[] data, NodeExistsPolicy policy)
       throws KeeperException, InterruptedException {
-    return putPersistentData(zPath, data, policy, ZooUtil.PUBLIC);
+    putPersistentData(zPath, data, policy, ZooUtil.PUBLIC);
   }
 
-  public boolean putPrivatePersistentData(String zPath, byte[] data, NodeExistsPolicy policy)
+  public void putPrivatePersistentData(String zPath, byte[] data, NodeExistsPolicy policy)
       throws KeeperException, InterruptedException {
-    return putPersistentData(zPath, data, policy, ZooUtil.PRIVATE);
+    putPersistentData(zPath, data, policy, ZooUtil.PRIVATE);
   }
 
-  public boolean putPersistentData(String zPath, byte[] data, NodeExistsPolicy policy,
-      List<ACL> acls) throws KeeperException, InterruptedException {
-    return putData(getZooKeeper(), getRetryFactory(), zPath, data, CreateMode.PERSISTENT, policy,
-        acls);
+  public void putPersistentData(String zPath, byte[] data, NodeExistsPolicy policy, List<ACL> acls)
+      throws KeeperException, InterruptedException {
+    putData(zPath, data, CreateMode.PERSISTENT, policy, acls);
   }
 
   public String putPersistentSequential(String zPath, byte[] data)
       throws KeeperException, InterruptedException {
-    final Retry retry = getRetryFactory().createRetry();
-    while (true) {
-      try {
-        return getZooKeeper().create(zPath, data, ZooUtil.PUBLIC, CreateMode.PERSISTENT_SEQUENTIAL);
-      } catch (KeeperException e) {
-        final Code c = e.code();
-        if (c == Code.CONNECTIONLOSS || c == Code.OPERATIONTIMEOUT || c == Code.SESSIONEXPIRED) {
-          retryOrThrow(retry, e);
-        } else {
-          throw e;
-        }
-      }
-
-      retry.waitForNextAttempt();
-    }
+    return retryLoop(
+        zk -> zk.create(zPath, data, ZooUtil.PUBLIC, CreateMode.PERSISTENT_SEQUENTIAL));
```

Comment:
Looking at this I wonder if on retry multiple sequential nodes could be created.  If so, it would be an existing issue.

Reply:
That's exactly the kind of observations I had hoped would come out of the simplification/refactor. :smiley_cat:

My guess is there's only a few places where this is called. Might look into that as follow-on.